subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RichVersionConstraintsResolveIntegrationTest.groovy

```diff
@@ -149,7 +151,89 @@ class StrictDependenciesResolveIntegrationTest extends AbstractModuleDependencyR
         fails ':checkDeps'
 
         then:
-        failure.assertHasCause('Cannot find a version of \'org:foo\' that satisfies the constraints: prefers 17, rejects ]17,), prefers 15, rejects ]15,)')
+        failure.assertHasCause("""Cannot find a version of 'org:foo' that satisfies the version constraints: 
+   Dependency path ':test:unspecified' --> 'org:foo' prefers '17', rejects ']17,)'
+   Dependency path ':test:unspecified' --> 'org:bar:1.0' --> 'org:foo' prefers '15', rejects ']15,)'""")
+
+    }
+
+    def "should fail during conflict resolution transitive dependency rejects"() {
+        given:
+        repository {
+            'org:foo' {
+                '1.0'()
+                '1.1'()
+            }
+            'org:bar:1.0' {
+                dependsOn(group: 'org', artifact: 'foo', version: '1.0', rejects: ['1.1']) // transitive dependency rejects version
+            }
+        }
+
+        buildFile << """
+            dependencies {
+                conf 'org:foo:1.1'
+                conf 'org:bar:1.0'
+            }           
+        """
+
+        when:
+        repositoryInteractions {
+            'org:foo:1.1' {
+                expectGetMetadata()
+            }
+            'org:bar:1.0' {
+                expectGetMetadata()
+            }
+        }
+        fails ':checkDeps'
+
+        then:
+        failure.assertHasCause("""Cannot find a version of 'org:foo' that satisfies the version constraints: 
+   Dependency path ':test:unspecified' --> 'org:foo' prefers '1.1'
+   Dependency path ':test:unspecified' --> 'org:bar:1.0' --> 'org:foo' prefers '1.0', rejects '1.1'""")
+    }
+
+    @Unroll
+    void "honors multiple rejections #rejects using dynamic versions using dependency notation #notation"() {
+        given:
+        repository {
+            (0..5).each {
+                "org:foo:1.$it"()
+            }
+            'org:bar:1.0' {
+                dependsOn(group: 'org', artifact: 'foo', version: '1.0', rejects: rejects) // transitive dependency rejects version
+            }
+        }
+
+        buildFile << """
+            dependencies {
+                conf 'org:foo:1.1'
+                conf 'org:bar:1.0'
+            }           
+        """
+
+        when:
+        repositoryInteractions {
+            'org:bar:1.0' {
+                expectGetMetadata()
+            }
+            'org:foo:1.1' {
+                expectGetMetadata()
+            }
+        }
+        fails ':checkDeps'
+
+        then:
+        failure.assertHasCause("""Cannot find a version of 'org:foo' that satisfies the version constraints: 
+   Dependency path ':test:unspecified' --> 'org:foo' prefers '1.1'
+   Dependency path ':test:unspecified' --> 'org:bar:1.0' --> 'org:foo' prefers '1.0', rejects any of "${rejects.collect {"'$it'"}.join(', ')}\"""")
+
+        where:
+        rejects << [
+            ['1.4', '1.5', '1.1'],
+            ['[1.1,)', '1.5'],
+            ['1.5', '[1.1, 1.3]', '1.4'],
```

Comment:
Needs coverage for `reject '+'`

Reply:
I added some: https://github.com/gradle/gradle/commit/03a16916f211d643f4f6a2fe8d553851d9b5f4ed